src/joyride/extension.cljs

```diff
@@ -30,8 +30,14 @@
      (when input
        (run-code input))))
   ([code]
-   (p/let [result (jsci/eval-string code)]
-     (utils/say-result result))))
+   (-> (p/let [result (jsci/eval-string code)]
+         (utils/say-result result)
+         result)
+       (p/catch (fn [e]

```

Comment:
Why is it necessary to transform the exception here and re-throw it?

Reply:
It was hard to consume from JavaScript. Maybe I did something wrong…

Reply:
New proposal for the data:

``` clojure
{:cause e
:ex-data (ex-data e)}
```
So the full exception is still preserved and accessible.

Reply:
The `js/Error` constructor throws away `:data`.

Reply:
And `:cause` gets hard to consume from JS. Maybe that's fine, but to me it made sense to use a js object there.

Reply:
I think using `:cause` for anything else than an exception is a bit unexpected?

Reply:
Other than that, I don't have any strong feelings. I wonder if the result should ever be parsed back in CLJS? Then it would be nice to preserve some of the original stuff, else it doesn't matter.

Reply:
I'm not very much in the know about JS exceptions... But since the only real hook we seem to have here is `:cause` maybe we can allow ourselves this, if we document it properly.

I do think most users will use this from CLJS... I just pushed a commit that tries to cater for both JS and CLJS user:

``` clojure
{:cause {:cljs-error e
         :data (ex-data e)}}
```

My testing RFC:

``` clojure
(comment
  (-> (p/let [r (vscode/commands.executeCommand "joyride.runCode" "null")]
        (def r r)
        (vscode/window.showInformationMessage (str r)))
      (p/catch (fn [e]
                 (def e e)
                 (vscode/window.showErrorMessage (.-message e)))))
  (-> e .-cause .-data)
  ;; => #js {:column 1, :file nil, :line 1, :phase "analysis", :type "error"}

  (-> e .-cause .-cljs-error ex-data)
  ;; => {:column 1, :file nil, :line 1, :phase "analysis", :type :sci/error}
  )
```

Reply:
> I do think most users will use this from CLJS

In this case I would opt to not wrap the exception at all? I'm not sure what problem is being solved here :-)

Reply:
Most users is not all users. 😄 If we don't wrap, then JS users can't get at the data. If we wrap, we can make the data reachable from both worlds.

Reply:
They can get all the data by evaluating another bit of joyride code?

Reply:
E.g. we don't convert the return value from run-code either to JS objects. It feels a bit pre-mature to do this.

Reply:
> They can get all the data by evaluating another bit of joyride code?

Not in any convenient way. If at all?

``` clojure
  (-> (p/let [r (vscode/commands.executeCommand "joyride.runCode" "null")]
        (vscode/window.showInformationMessage (str r)))
      (p/catch (fn [e]
                 (p/let [js-error-data (vscode/commands.executeCommand "joyride.runCode" can't-construct-this-arg-I-think)
                         ]
                   use-js-error-data ...))))
```

> E.g. we don't convert the return value from run-code either to JS objects.

Good catch! We probably should. It's a fault of me just having testing this with primitives.

Reply:
🤦 No, I don't think we should do that at all. Too much magic. When people want to have a primitive JS result, they can convert it as part of the program.

Reply:
One way you can do this manually is to evaluate `clj->js` and then do the conversion yourself using that function.

Reply:
I've pushed some commits removing the wrapping of the exception, and adding some documentation around. Please have a look!